README.md

```diff
@@ -41,92 +41,145 @@ simplest way.
+
+            return $this->redirectToRoute('app_register');
+        }
+
+        // @TODO Change the redirect on success and handle or remove the flash message in your templates
+        $this->addFlash('success', 'Your email address has been verified.');
+
+        return $this->redirectToRoute('app_register');
     }
 }
 ```
 
-Once the user has received their email and clicked on the link, the `RegistrationController`
-would then validate the signed URL in following method:
+This uses an `EmailVerifier` class that you should also add to your app::
 
 ```php
-// RegistrationController.php
+// src/Security/EmailVerifier.php
+namespace App\Security;
 
+use Doctrine\ORM\EntityManagerInterface;
+use Symfony\Bridge\Twig\Mime\TemplatedEmail;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\Mailer\MailerInterface;
+use Symfony\Component\Security\Core\User\UserInterface;
 use SymfonyCasts\Bundle\VerifyEmail\Exception\VerifyEmailExceptionInterface;
-// ...
+use SymfonyCasts\Bundle\VerifyEmail\VerifyEmailHelperInterface;
 
-class RegistrationController extends AbstractController
+class EmailVerifier
 {
-    // ...
-    /**
-     * @Route("/verify", name="registration_confirmation_route")
-     */
-    public function verifyUserEmail(Request $request): Response
+    public function __construct(
+        private VerifyEmailHelperInterface $verifyEmailHelper,
+        private MailerInterface $mailer,
+        private EntityManagerInterface $entityManager
+    ) {
+    }
+
+    public function sendEmailConfirmation(string $verifyEmailRouteName, UserInterface $user, TemplatedEmail $email): void
     {
-        $this->denyAccessUnlessGranted('IS_AUTHENTICATED_FULLY');
-        $user = $this->getUser();
+        $signatureComponents = $this->verifyEmailHelper->generateSignature(
+            $verifyEmailRouteName,
+            $user->getId(),
+            $user->getEmail()
+        );
 
-        // Do not get the User's Id or Email Address from the Request object
-        try {
-            $this->verifyEmailHelper->validateEmailConfirmation($request->getUri(), $user->getId(), $user->getEmail());
-        } catch (VerifyEmailExceptionInterface $e) {
-            $this->addFlash('verify_email_error', $e->getReason());
+        $context = $email->getContext();
+        $context['signedUrl'] = $signatureComponents->getSignedUrl();
+        $context['expiresAtMessageKey'] = $signatureComponents->getExpirationMessageKey();
+        $context['expiresAtMessageData'] = $signatureComponents->getExpirationMessageData();
 
-            return $this->redirectToRoute('app_register');
-        }
+        $email->context($context);
 
-        // Mark your user as verified. e.g. switch a User::verified property to true
+        $this->mailer->send($email);
+    }
+
+    /**
+     * @throws VerifyEmailExceptionInterface
+     */
+    public function handleEmailConfirmation(Request $request, UserInterface $user): void
```

Comment:
Hm, IIRC `UserInterface` does not have `getId()`/`getEmail()` methods, so maybe we should typhint with `App\Entity\User` instead?

Reply:
We chewed on this back when we were first drafting out the code for this. IIRC, because we don't know if the user object _is_ called `User::class` - we went with the safe bet that whatever the "user" is, it will more likely than not implement `UserInterface`. Going to merge as is because this is what we generate in MakerBundle. But we can definitely adjust as needed.